File: Tennis.Logic/Set.cs
using System;

namespace Tennis.Logic
{
	public class Set
	{
		int sideOneScore;
		int sideTwoScore;
		SetState state;
		
		public Set()
		{
			state = SetState.Playing;
		}
		
		public void WinGame(Func<Side, Side> scoring)
		{
			var scoringSide = scoring(Side.None);
			AdvanceState(scoringSide);
		}
		
		public int SideOneScore
		{
			get { return sideOneScore; }
		}
		
		public int SideTwoScore
		{
			get { return sideTwoScore; }
		}
		
		public SetState State 
		{ 
			get { return state; }
		}
		
		public string PrintScore()
		{
			return sideOneScore.ToString() + " - " + sideTwoScore.ToString();
		}
		
		private void AdvanceState(Side side)
		{
			if (side == Side.One)
			{
				sideOneScore++;
			}
			
			if (side == Side.Two)
			{
				sideTwoScore++;
			}
			
			var diff = sideOneScore - sideTwoScore;
			if (Math.Abs(diff) >= 2)
			{
				if (sideOneScore >= 6 || sideTwoScore >= 6)
				{
					if (diff > 0)
					{
						state = SetState.SetWonBySideOne;
					}
					else
					{
						state = SetState.SetWonBySideTwo;
					}
				}
			}
		}
	}
}

File: Tennis.FSharp.Logic/Set.fs
namespace Tennis.FSharp.Logic

    open System
    open FSharpx

    type Set() =
        let mutable setState = SetState.Playing
        let mutable sideOneScore = 0
        let mutable sideTwoScore = 0

        member this.SideOneScore
            with get () = sideOneScore

        member this.SideTwoScore
            with get () = sideTwoScore

        member this.State
            with get () = setState

        member private this.IsWinByTwo =
            if sideOneScore >= 5 && sideTwoScore >= 5 then
                true
            else
                false

        member private this.DetermineWinner =
            if this.IsWinByTwo then
                if sideOneScore > sideTwoScore && sideOneScore - sideTwoScore >= 2 then
                    setState <- SetState.SetWonBySideOne
                elif sideTwoScore > sideOneScore && sideTwoScore - sideOneScore >= 2 then
                    setState <- SetState.SetWonBySideTwo
            else
                if sideOneScore >= 6 then
                    setState <- SetState.SetWonBySideOne
                elif sideTwoScore >= 6 then
                    setState <- SetState.SetWonBySideTwo

        member this.WinGame(f: Func<_,_>) =
            let func = FSharpFunc.FromFunc f
            let winner = func Side.None
            
            setState <- SetState.Playing
            
            if winner = Side.One then
                sideOneScore <- sideOneScore + 1
            else
                sideTwoScore <- sideTwoScore + 1

            this.DetermineWinner

        member this.PrintScore() =
            sideOneScore.ToString() + " - " + sideTwoScore.ToString()